neuralmonkey/tf_manager.py

```diff
@@ -0,0 +1,178 @@
+    """
+
+    def __init__(self, num_sessions, num_threads, variable_files=None,
+                 gpu_allow_growth=True):
+        """Initialize a TensorflowManager.
+
+        At this moment the graph must already exist. This method initializes
+        required number of TensorFlow sessions and initializes them with
+        provided variable files if they are provided.
+
+        Args:
+            num_sessions: Number of sessions to be initialized.
+            num_threads: Number of threads sessions will run in.
+            variable_files: List of variable files.
+        """
+
+        session_cfg = tf.ConfigProto()
+        session_cfg.inter_op_parallelism_threads = num_threads
+        session_cfg.intra_op_parallelism_threads = num_threads
+        session_cfg.allow_soft_placement = True # needed for multiple GPUs
+        # pylint: disable=no-member
+        session_cfg.gpu_options.allow_growth = gpu_allow_growth
+
+        self.sessions = [tf.Session(config=session_cfg)
+                         for _ in range(num_sessions)]
+        init_op = tf.initialize_all_variables()
+        for sess in self.sessions:
+            sess.run(init_op)
+        self.saver = tf.train.Saver()
+
+        if variable_files:
+            if len(variable_files) != num_sessions:
+                raise Exception(("The number of provided variable files ({}) "
+                                 "is different than a number sessions ({})")
+                                .format(len(variable_files), num_sessions))
+            self.restore(variable_files)
+
+    # pylint: disable=too-many-locals
+    def execute(self,
+                dataset: Dataset,
+                execution_scripts,
+                train=False,
+                summaries=True,
+                batch_size=None) -> List[ExecutionResult]:
+        if batch_size is None:
+            batch_size = len(dataset)
+        batched_dataset = dataset.batch_dataset(batch_size)
+
+        batch_results = [[] for _ in
+                         execution_scripts]  # type: List[List[ExecutionResult]]
+        for batch in batched_dataset:
+            executables = [s.get_executable(train=train, summaries=summaries)
+                           for s in execution_scripts]
+            while not all(ex.result is not None for ex in executables):
+                all_feedables = set()   # type: Set[Any]
+                all_tensors_to_execute = []  # type: List[tf.Tensor]
+                additional_feed_dicts = []
+                tensor_list_lengths = []  # type: List[int]
+
+                for executable in executables:
+                    if executable.result is None:
+                        (feedables,
+                         tensors_to_execute,
+                         add_feed_dict) = executable.next_to_execute()
+                        all_feedables = all_feedables.union(feedables)
+                        all_tensors_to_execute.extend(tensors_to_execute)
```

Comment:
Tady vidim jedinou záškobrtku v tom způsobu fetchování popsanym v runneru.. Ale asi bude stačit tady vyměnit extend za update a pak asi ještě šéfovat, aby nekolidovaly klíče v těch dictionaries.. Ale on by vlastně ``all_tensors_to_execute`` mohl být dict indexovanej runner objektama a pak tohle by ani nebylo update, ale jen přiřazení.

A úplně se ten kód zbaví tý režie okolo, jako ``tensor_list_lengths``, nebo k čemu je ta proměnná.